d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java

```diff
@@ -439,117 +439,112 @@ public void testClusterInfoProviderUnregisterClusterListener()
+                                        new HashMap<>(), new HashMap<>()));
+
+    doAnswer(invocation ->
+             {
+               Thread.sleep(10);
+               return null;
+             }).when(state).listenToCluster(any(), any());
+
+    SimpleLoadBalancer loadBalancer = spy(new SimpleLoadBalancer(state, 1, TimeUnit.MILLISECONDS, _d2Executor));
+    FutureCallback<Pair<ClusterProperties, UriProperties>> callback = spy(new FutureCallback<>());
+    // case1: listenToCluster timeout, and simpleLoadBalancer not hit the cache value
     loadBalancer.getLoadBalancedClusterAndUriProperties(CLUSTER1_NAME, callback);
+    try
+    {
+      callback.get();
+    }
+    catch (ExecutionException e)
+    {
+      Assert.assertTrue(e.getCause() instanceof ServiceUnavailableException);
+    }
+    verify(loadBalancer).handleTimeoutFromGetClusterAndUriProperties(eq(CLUSTER1_NAME), eq(callback));
+    verify(callback).onError(any(ServiceUnavailableException.class));
 
-    // case2: getLoadBalancedClusterAndUriProperties timeout, and simpleLoadBalancer hit the cache value from state
+    // case2: listenToCluster timeout, and simpleLoadBalancer hit the cache value from state
     LoadBalancerStateItem<ClusterProperties> clusterItem = new LoadBalancerStateItem<>(CLUSTER_PROPERTIES, 1, 1);
     LoadBalancerStateItem<UriProperties> uriItem = new LoadBalancerStateItem<>(URI_PROPERTIES, 1, 1);
-    LoadBalancerTestState state  = spy(new LoadBalancerTestState());
-    state.listenToCluster = false;
-    state.listenToService = false;
     when(state.getClusterProperties(CLUSTER1_NAME)).thenReturn(clusterItem);
     when(state.getUriProperties(CLUSTER1_NAME)).thenReturn(uriItem);
-    loadBalancer = spy(setupLoadBalancer(state, serviceRegistry, clusterRegistry, uriRegistry));
-    callback = new Callback<Pair<ClusterProperties, UriProperties>>()
-    {
-      @Override
-      public void onError(Throwable e)
-      {
-        throw new RuntimeException("onError should not be called");
-      }
-
-      @Override
-      public void onSuccess(Pair<ClusterProperties, UriProperties> result)
-      {
-        assertEquals(result.getLeft(), CLUSTER_PROPERTIES);
-        assertEquals(result.getRight(), URI_PROPERTIES);
-      }
-    };
-
+    callback = spy(new FutureCallback<>());
     loadBalancer.getLoadBalancedClusterAndUriProperties(CLUSTER1_NAME, callback);
+    callback.get();
+    verify(callback).onSuccess(eq(Pair.of(CLUSTER_PROPERTIES, URI_PROPERTIES)));
+
 
     // case3: getLoadBalancedClusterAndUriProperties without timeout
-    loadBalancer = setupLoadBalancer(null, serviceRegistry, clusterRegistry, uriRegistry);
+    state =
+        spy(new SimpleLoadBalancerState(new SynchronousExecutorService(), uriRegistry, clusterRegistry, serviceRegistry,
+                                        new HashMap<>(), new HashMap<>()));
+    loadBalancer = spy(new SimpleLoadBalancer(state, 5, TimeUnit.SECONDS, _d2Executor));
```

Comment:
these state, load balancer, and the callback below are not seem to be spied for mocking any method's result (like mocking with doReturn, thenReturn, doAnswer, etc? they could just use real objects, right? 

Same applied to some other objects in other tests.

Reply:
Hi, @bohhyang 
#1. state needs spy for
```java
 doAnswer(invocation ->
           {
             Thread.sleep(10);
             return null;
           }).when(state).listenToService(any(), any());
```

#2 If I don't use `spy` or `mock` for `callback` and `loadBalancer`. The test will be failed for
```java
org.mockito.exceptions.misusing.NotAMockException: 
Argument passed to verify() is of type FutureCallback and is not a mock!
Make sure you place the parenthesis correctly!
See the examples of correct verifications:
    verify(mock).someMethod();
    verify(mock, times(10)).someMethod();
    verify(mock, atLeastOnce()).someMethod();
```